modules/dxp/apps/multi-factor-authentication/multi-factor-authentication-fido2-web/src/main/java/com/liferay/multi/factor/authentication/fido2/web/internal/portlet/action/MFAFIDO2UserAccountSetupMVCActionCommand.java

```diff
@@ -0,0 +1,128 @@
+
+import com.liferay.multi.factor.authentication.fido2.web.internal.checker.FIDO2BrowserSetupMFAChecker;
+import com.liferay.multi.factor.authentication.spi.checker.setup.SetupMFAChecker;
+import com.liferay.osgi.service.tracker.collections.map.PropertyServiceReferenceMapper;
+import com.liferay.osgi.service.tracker.collections.map.ServiceTrackerMap;
+import com.liferay.osgi.service.tracker.collections.map.ServiceTrackerMapFactory;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.portlet.bridges.mvc.BaseMVCActionCommand;
+import com.liferay.portal.kernel.portlet.bridges.mvc.MVCActionCommand;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.ParamUtil;
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.users.admin.constants.UsersAdminPortletKeys;
+
+import javax.portlet.ActionRequest;
+import javax.portlet.ActionResponse;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Arthur Chan
+ */
+@Component(
+	property = {
+		"javax.portlet.name=" + UsersAdminPortletKeys.MY_ACCOUNT,
+		"javax.portlet.name=" + UsersAdminPortletKeys.USERS_ADMIN,
+		"mvc.command.name=/my_account/setup_mfa/fido2/remove_authenticator"
+	},
+	service = MVCActionCommand.class
+)
+public class MFAFIDO2UserAccountSetupMVCActionCommand
+	extends BaseMVCActionCommand {
+
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_mfaCheckerServiceTrackerMap =
+			ServiceTrackerMapFactory.openSingleValueMap(
+				bundleContext, SetupMFAChecker.class, "(service.id=*)",
+				new PropertyServiceReferenceMapper<>("service.id"));
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		_mfaCheckerServiceTrackerMap.close();
+	}
+
+	@Override
+	protected void doProcessAction(
+			ActionRequest actionRequest, ActionResponse actionResponse)
+		throws Exception {
+
+		long setupMFACheckerServiceId = ParamUtil.getLong(
+			actionRequest, "setupMFACheckerServiceId");
+
+		SetupMFAChecker setupMFAChecker =
+			_mfaCheckerServiceTrackerMap.getService(setupMFACheckerServiceId);
+
+		if (setupMFAChecker == null) {
+			_log.error(
+				"Unable to get setup MFA checker " + setupMFACheckerServiceId);
+
+			return;
+		}
+
+		FIDO2BrowserSetupMFAChecker fido2BrowserSetupMFAChecker = null;
+
+		try {
+			fido2BrowserSetupMFAChecker =
+				(FIDO2BrowserSetupMFAChecker)setupMFAChecker;
+		}
+		catch (ClassCastException classCastException) {
+			_log.error(classCastException.getMessage());
+
+			return;
+		}
+
+		long authenticatorId = ParamUtil.getLong(
+			actionRequest, "authenticatorId");
```

Comment:
it looks like we need to pair this with the user or at least the company, we can't trust the data from the request.